ethpm/tools/checker.py

```diff
@@ -0,0 +1,262 @@
+    """should not include abstract contracts in the contract types section of a release.""",
+    "abi_missing": """Contract type: {0} is missing an abi field, which is essential for using """
+    """this package.""",
+    "deployment_bytecode_missing": """Contract type: {0} is missing a deployment_bytecode field,"""
+    """ which is essential for using this package.""",
+    "contract_type_subfield_missing": """Contract type: {0} is missing a contract_type field, """
+    """which is essential if an alias is being used to namespace this contract type.""",
+    "runtime_bytecode_missing": "Contract type: {0} is missing a runtime_bytecode field.",
+    "bytecode_subfield_missing": """Contract type: {0} is missing a required bytecode subfield """
+    """in its {1} bytecode object.""",
+    "natspec_missing": "Contract type: {0} is missing a natspec field.",
+    "compiler_missing": "Contract type: {0} is missing a compiler field.",
+}
+
+
+#
+# Validation
+#
+
+
+def check_manifest(manifest):
+    generate_warnings = (
+        check_manifest_version(manifest),
+        check_package_name(manifest),
+        check_version(manifest),
+        check_meta(manifest),
+        check_sources(manifest),
+        check_contract_types(manifest),
+    )
+    return build({}, *generate_warnings)
+
+
+#
+# Required fields
+#
+
+
+@cytoolz.curry
+def check_manifest_version(manifest, warnings):
+    if "manifest_version" not in manifest or not manifest["manifest_version"]:
+        return assoc(warnings, "manifest_version", WARNINGS["manifest_version_missing"])
+    if manifest["manifest_version"] != "2":
+        return assoc(warnings, "manifest_version", WARNINGS["manifest_version_invalid"])
+    return warnings
+
+
+@cytoolz.curry
+def check_package_name(manifest, warnings):
+    if "package_name" not in manifest or not manifest["package_name"]:
+        return assoc(warnings, "package_name", WARNINGS["package_name_missing"])
+    if not bool(re.match(PACKAGE_NAME_REGEX, manifest["package_name"])):
+        return assoc(warnings, "package_name", WARNINGS["package_name_invalid"])
+    return warnings
+
+
+@cytoolz.curry
+def check_version(manifest, warnings):
+    if "version" not in manifest or not manifest["version"]:
+        return assoc(warnings, "version", WARNINGS["version_missing"])
+    return warnings
+
+
+#
+# Meta fields
+#
+
+
+@cytoolz.curry
+def check_meta(manifest, warnings):
+    if "meta" not in manifest or not manifest["meta"]:
+        return assoc(warnings, "meta", WARNINGS["meta_missing"])
+    meta_validation = (
+        check_authors(manifest["meta"]),
+        check_license(manifest["meta"]),
+        check_description(manifest["meta"]),
+        check_keywords(manifest["meta"]),
+        check_links(manifest["meta"]),
+    )
+    return build(warnings, *meta_validation)
+
+
+@cytoolz.curry
+def check_authors(meta, warnings):
+    if "authors" not in meta:
+        return assoc(warnings, "authors", WARNINGS["authors_missing"])
```

Comment:
Probably worth namespacing the string as `"meta.authors"` so that the `warnings` dictionary keys everything by the dot-separated-path that it would be looked up from the manifest.